test/core/iomgr/tcp_client_posix_test.cc

```diff
@@ -207,27 +211,82 @@ void test_fails(void) {
 
 void test_connect_cancellation_succeeds(void) {
   gpr_log(GPR_ERROR, "---- starting test_connect_cancellation_succeeds() ----");
+  auto target_ipv6_addr_uri = *grpc_core::URI::Parse(absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die())));
+  auto target_ipv4_addr_uri = *grpc_core::URI::Parse(absl::StrCat(
+      "ipv4:127.0.0.1:", std::to_string(grpc_pick_unused_port_or_die())));
   grpc_resolved_address resolved_addr;
-  struct sockaddr_in* addr =
-      reinterpret_cast<struct sockaddr_in*>(resolved_addr.addr);
+  ASSERT_TRUE(grpc_parse_uri(target_ipv6_addr_uri, &resolved_addr));
   int svr_fd;
   grpc_closure done;
   grpc_core::ExecCtx exec_ctx;
-
-  memset(&resolved_addr, 0, sizeof(resolved_addr));
-  resolved_addr.len = static_cast<socklen_t>(sizeof(struct sockaddr_in));
-  addr->sin_family = AF_INET;
-
+  bool tried_ipv4 = false;
   /* create a phony server */
-  svr_fd = socket(AF_INET, SOCK_STREAM, 0);
-  ASSERT_GE(svr_fd, 0);
-  ASSERT_EQ(bind(svr_fd, (struct sockaddr*)addr, (socklen_t)resolved_addr.len),
-            0);
+  svr_fd = socket(AF_INET6, SOCK_STREAM, 0);
+  if (svr_fd < 0 ||
+      bind(svr_fd, reinterpret_cast<sockaddr*>(resolved_addr.addr),
+           resolved_addr.len) != 0) {
+    // Try ipv4
+    ASSERT_TRUE(grpc_parse_uri(target_ipv4_addr_uri, &resolved_addr));
+    svr_fd = socket(AF_INET, SOCK_STREAM, 0);
+    tried_ipv4 = true;
+    if (svr_fd < 0 ||
+        bind(svr_fd, reinterpret_cast<sockaddr*>(resolved_addr.addr),
+             resolved_addr.len) != 0) {
+      gpr_log(GPR_ERROR,
+              "Skipping test. Failed to create a phony server bound to ipv6 or "
+              "ipv4 address");
+      return;
+    }
+  }
+
   ASSERT_EQ(listen(svr_fd, 1), 0);
 
+  std::vector<int> client_sockets;
+  bool create_more_client_connections = true;
+  // Create and connect client sockets until the connection attempt times out.
+  // Even if the backlog specified to listen is 1, the kernel continues to
+  // accept a certain number of SYN packets before dropping them. This loop
+  // attempts to identify the number of new connection attempts that will
+  // be allowed by the kernel before any subsequent connection attempts
+  // become pending indefinitely.
+  while (create_more_client_connections) {
+    const int kOne = 1;
+    int client_socket = socket(tried_ipv4 ? AF_INET : AF_INET6, SOCK_STREAM, 0);
+    ASSERT_GT(client_socket, 0);
+    setsockopt(client_socket, SOL_SOCKET, SO_REUSEADDR, &kOne, sizeof(kOne));
+    // Make fd non-blocking.
+    int flags = fcntl(client_socket, F_GETFL, 0);
+    EXPECT_EQ(fcntl(client_socket, F_SETFL, flags | O_NONBLOCK), 0);
```

Comment:
It looks like it should be `ASSERT_EQ` instead.

Reply:
Fixed.